File: Mathematical_Algorithms/Java/miller_rabin_test.java
import java.math.BigInteger;
import java.util.Random;

public class miller_rabin_test {

    // Hàm tính (a^d) % n
    private static BigInteger power(BigInteger base, BigInteger exp, BigInteger mod) {
        BigInteger res = BigInteger.ONE;
        base = base.mod(mod);
        while (exp.compareTo(BigInteger.ZERO) > 0) {
            if (exp.testBit(0)) { // Nếu exp là lẻ
                res = res.multiply(base).mod(mod);
            }
            base = base.multiply(base).mod(mod);
            exp = exp.shiftRight(1); // exp = exp / 2
        }
        return res;
    }

    // Hàm trợ giúp cho Miller-Rabin, kiểm tra một cơ số a
    private static boolean millerRabinTest(BigInteger d, BigInteger n) {
        Random rand = new Random();
        // Chọn một số ngẫu nhiên trong [2, n-2]
        BigInteger a = new BigInteger(n.bitLength(), rand);
        while (a.compareTo(BigInteger.TWO) < 0 || a.compareTo(n.subtract(BigInteger.TWO)) > 0) {
            a = new BigInteger(n.bitLength(), rand);
        }

        BigInteger x = power(a, d, n);

        if (x.equals(BigInteger.ONE) || x.equals(n.subtract(BigInteger.ONE))) {
            return true;
        }

        // Bình phương x lặp đi lặp lại
        while (d.compareTo(n.subtract(BigInteger.ONE)) != 0) {
            x = x.multiply(x).mod(n);
            d = d.multiply(BigInteger.TWO);

            if (x.equals(BigInteger.ONE)) {
                return false;
            }
            if (x.equals(n.subtract(BigInteger.ONE))) {
                return true;
            }
        }
        return false;
    }

    // Kiểm tra tính nguyên tố của n bằng Miller-Rabin k lần
    public static boolean isPrimeMillerRabin(BigInteger n, int k) {
        // Trường hợp đặc biệt
        if (n.compareTo(BigInteger.ONE) <= 0 || n.equals(BigInteger.valueOf(4))) {
            return false;
        }
        if (n.compareTo(BigInteger.valueOf(3)) <= 0) {
            return true;
        }

        // Tìm d sao cho n-1 = 2^s * d
        BigInteger d = n.subtract(BigInteger.ONE);
        while (d.mod(BigInteger.TWO).equals(BigInteger.ZERO)) {
            d = d.divide(BigInteger.TWO);
        }

        // Lặp lại k lần
        for (int i = 0; i < k; i++) {
            if (!millerRabinTest(d, n)) {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {
        System.out.println("13 là số nguyên tố: " + isPrimeMillerRabin(BigInteger.valueOf(13), 4));
        System.out.println("15 là số nguyên tố: " + isPrimeMillerRabin(BigInteger.valueOf(15), 4));
        System.out.println("97 là số nguyên tố: " + isPrimeMillerRabin(BigInteger.valueOf(97), 4));
        System.out.println("561 là số nguyên tố: " + isPrimeMillerRabin(BigInteger.valueOf(561), 4)); // 561 là số Carmichael, Miller-Rabin sẽ phát hiện ra nó là hợp số

        // Kiểm tra một số lớn
        BigInteger largePrimeCandidate = new BigInteger("1000000007");
        System.out.println(largePrimeCandidate + " là số nguyên tố: " + isPrimeMillerRabin(largePrimeCandidate, 10));

        BigInteger largeCompositeCandidate = new BigInteger("1000000001"); // 1000000001 = 101 * 9900990.10
        System.out.println(largeCompositeCandidate + " là số nguyên tố: " + isPrimeMillerRabin(largeCompositeCandidate, 10));
    }
}

File: Mathematical_Algorithms/JavaScript/miller_rabin_test.js
function power(base, exp, mod) {
    let res = 1n; // Sử dụng BigInt
    base = base % mod;
    while (exp > 0n) {
        if (exp % 2n === 1n) {
            res = (res * base) % mod;
        }
        base = (base * base) % mod;
        exp = exp / 2n;
    }
    return res;
}

function millerRabinTest(d, n) {
    // Chọn một số ngẫu nhiên trong [2, n-2]
    const a = BigInt(Math.floor(Math.random() * (Number(n - 2n) - 2)) + 2);
    let x = power(a, d, n);

    if (x === 1n || x === n - 1n) {
        return true;
    }

    // Bình phương x lặp đi lặp lại
    while (d !== n - 1n) {
        x = (x * x) % n;
        d *= 2n;

        if (x === 1n) {
            return false;
        }
        if (x === n - 1n) {
            return true;
        }
    }
    return false;
}

function isPrimeMillerRabin(n, k = 4) {
    n = BigInt(n);

    // Trường hợp đặc biệt
    if (n <= 1n || n === 4n) {
        return false;
    }
    if (n <= 3n) {
        return true;
    }

    // Tìm d sao cho n-1 = 2^s * d
    let d = n - 1n;
    while (d % 2n === 0n) {
        d /= 2n;
    }

    // Lặp lại k lần
    for (let i = 0; i < k; i++) {
        if (!millerRabinTest(d, n)) {
            return false;
        }
    }
    return true;
}

// Ví dụ sử dụng:
console.log(`13 là số nguyên tố: ${isPrimeMillerRabin(13)}`);
console.log(`15 là số nguyên tố: ${isPrimeMillerRabin(15)}`);
console.log(`97 là số nguyên tố: ${isPrimeMillerRabin(97)}`);
console.log(`561 là số nguyên tố: ${isPrimeMillerRabin(561)}`); // 561 là số Carmichael, Miller-Rabin sẽ phát hiện ra nó là hợp số

// Kiểm tra một số lớn
const largePrimeCandidate = 1000000007n;
console.log(`${largePrimeCandidate} là số nguyên tố: ${isPrimeMillerRabin(largePrimeCandidate, 10)}`);

const largeCompositeCandidate = 1000000001n; // 1000000001 = 101 * 9900990.10
console.log(`${largeCompositeCandidate} là số nguyên tố: ${isPrimeMillerRabin(largeCompositeCandidate, 10)}`);
